arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,315 @@
+  vil_image_view< PixType >
+  calculate_aligned_edges( vil_image_view< PixType > const& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j );
+
+  template < typename InputType, typename OutputType >
+  vil_image_view< OutputType >
+  nonmax_suppression( vil_image_view< InputType > const& grad_i,
+                      vil_image_view< InputType > const& grad_j );
+
+  template < typename pix_t > vil_image_view< pix_t >
+  filter( vil_image_view< pix_t > const& input_image );
+
+  aligned_edge_detection* p;
+  // Internal parameters/settings
+  float threshold = 10;
+  bool produce_joint_output = true;
+  double smoothing_sigma = 1.3;
+  unsigned smoothing_half_step = 2;
+};
+
+// ----------------------------------------------------------------------------
+// Perform NMS on the input gradient images in horizontal and vert directions
+// only
+template < typename InputType, typename OutputType >
+vil_image_view< OutputType >
+aligned_edge_detection::priv
+::nonmax_suppression( vil_image_view< InputType > const& grad_i,
+                      vil_image_view< InputType > const& grad_j )
+{
+  if( grad_i.ni() != grad_j.ni() || grad_i.nj() != grad_j.nj() )
+  {
+    LOG_ERROR(
+      p->logger(),
+      "Input gradient image dimensions must be equivalent" );
+  }
+
+  auto const ni = grad_i.ni();
+  auto const nj = grad_i.nj();
+
+  vil_image_view< OutputType > output( ni, nj, 2 );
+  output.fill( 0 );
+
+  // Perform non-maximum suppression
+  for( decltype(+nj) j = 1; j < nj - 1; j++ )
+  {
+    for( decltype(+ni) i = 1; i < ni - 1; i++ )
+    {
+      const InputType val_i = grad_i( i, j );
+      const InputType val_j = grad_j( i, j );
+
+      if( val_i > threshold )
+      {
+        if( val_i >= grad_i( i - 1, j ) && val_i >= grad_i( i + 1, j ) )
+        {
+          output( i, j, 0 ) = static_cast< OutputType >( val_i );
+        }
+      }
+      if( val_j > threshold )
+      {
+        if( val_j >= grad_j( i, j - 1 ) && val_j >= grad_j( i, j + 1 ) )
+        {
+          output( i, j, 1 ) = static_cast< OutputType >( val_j );
+        }
+      }
+    }
+  }
+  return output;
+}
+
+// ----------------------------------------------------------------------------
+// Calculate potential edges
+template < typename PixType, typename GradientType >
+vil_image_view< PixType >
+aligned_edge_detection::priv
+::calculate_aligned_edges( vil_image_view< PixType > const& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j )
+{
+  // Calculate sobel approx in x/y directions
+  vil_sobel_3x3( input, grad_i, grad_j );
+
+  // Take absolute value of gradients
+  vil_transform< GradientType, GradientType( GradientType ) >(
+    grad_i, std::abs );
+  vil_transform< GradientType, GradientType( GradientType ) >(
+    grad_j, std::abs );
```

Comment:
Okay... mostly because trying to read this is making me go "WTH?"...
```suggestion
  vil_transform( grad_i, std::abs< GradientType > );
  vil_transform( grad_j, std::abs< GradientType > );
```

Alternatively:
```suggestion
  vil_transform< GradientType, GradientType (*)( GradientType ) >(
    grad_i, std::abs );
  vil_transform< GradientType, GradientType (*)( GradientType ) >(
    grad_j, std::abs );
```
...but the former is more concise. (The problem is the second template argument looks like a nonsense object construction rather than a function type.)